### ISSUE TITLE
RFC: Deprecate Unpaused Main Looper

### ISSUE THREAD

**Issue by charlesmunger** [robolectric/robolectric#3369]
# RFC: Deprecate Unpaused Main Looper
## Background
One major difference between Robolectric and android emulator tests is that in robolectric, the test thread is the same as the UI thread. This makes it easy to make tests deterministic, and it removes the needs for lots of synchronization between the test thread and the UI thread. However, since the UI thread is an event loop, it means that runnables queued on the UI thread have to be manually run. In addition to the syntactic annoyance of draining the queue after almost every method call in a test, if test authors forgot to drain the queue at the end of the test they may miss a potential crash. Very early on in robolectric’s development, a mode was added that made this easier. If you have a method like this:

```
void doFoo() {
   postOnUiThread(this::doBar);
   doBaz();
}
```
With the unpaused main looper behavior, you can write
```
doFoo();
```
Instead of 
```
doFoo();
ShadowLooper.runUiThreadTasks();
```
However, doBar() will be executed inline from postOnUiThread - so it’ll run before baz(), even though on a real android device, doBaz() will always run before doBar(). The default behavior of Robolectric is the unpaused behavior. This can be toggled at runtime via the existing API of `ShadowLooper.pauseMainLooper()`.

## Cost
Like many other large companies, Google uses a monorepo. When we update the Support Library in our repo, many tests break. A common cause of these breakages is that tests use the unpaused looper behavior and reentrancy isn’t permitted with the use of some API. For example, FragmentManager (reasonably) doesn’t permit reentrantly calling executePendingTransactions. Code that’s correct on real android causes test failures in robolectric, because robolectric runs tasks inline, in violation of the actual API contract. Of all the upgrades we’ve done of the support library Google’s monorepo in the last 6 months, every single one triggered at least one test failure due to unpaused loopers, where, the code under test was actually correct.

There’s also the cost of running code in tests differently from how it runs in production. For example, a common cause of crashes in Android code is posting a runnable to the UI thread that references an Activity, View, or Fragment. If the host Activity is destroyed before that runnable runs, it’ll interact with a destroyed view hierarchy. Running tasks inline hides this source of crashes. 
## Proposal
We’d like to some day flip the default value of pausing the UI thread looper. We’d introduce a new API in RobolectricTestRunner that decided whether pausing the main looper was permitted, and another to specify the default state of the looper. This way, teams could build a list of existing tests that depend on the paused looper behavior, and make the default paused for new tests. Gradually, they could make unpausing the main looper throw an exception for tests not on the list. Over time, the list would shrink. In a far-off future release of Robolectric, the global default could flip from unpaused to paused, with a natural migration path for tests depending on the unpaused behavior. 

### Orphaned Tasks
One danger of migrating tests from running tasks inline to queuing them is that code like `post(() -> assertThat(someCondition).isTrue())` may not be run if the developer does not flush the queue before the test ends. We can fix this by either having robolectric flush the queue at the end of each test, or by failing the test if there are still runnables left in the queue when the test completes. I suspect that the latter option would interact better with junit `@Rule`s.

**Comment by hoisie** [comment]
Just a few thoughts on this:

1) Although this is definitely an issue with the support library. I'd be curious how widespread this problem is.
2) It may be possible to craft support library shadows that fixes the issues with reentrancy, e.g https://github.com/robolectric/robolectric/pull/3167. This would obviously require continued maintenance as the support library evolves.
3) By defaulting to paused, wouldn't that burden developers writing simple tests such as "setting up an activity and clicking on a button" to have to manually call 'runUiThreadTasks' for things like Activity setup?
4) I'm curious if there's anything we can do at the instrumentation level to address this. For example, during calls to Android methods, we could keep track of a 'deferred' list of tasks that are posted to the UI thread, and run them after the method intercept is over. This probably wouldn't be possible to do for user code, though.

**Comment by charlesmunger** [comment]
1. The support library includes a lot of things - I've seen it with simply throwing exceptions in fragment transactions, but I've also seen it produce cryptic failures like this one:
```
java.lang.NullPointerException
        at android.support.v4.app.Fragment.onDestroy(Fragment.java:1507)
        at android.support.v4.app.Fragment.performDestroy(Fragment.java:2426)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1437)
        at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1523)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1598)
        at android.support.v4.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2145)
        at android.support.v4.app.FragmentManagerImpl.optimizeAndExecuteOps(FragmentManager.java:2092)
        at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1998)
        at android.support.v4.app.FragmentManagerImpl$1.run(FragmentManager.java:709)
        at android.os.Handler.handleCallback(Handler.java:739)
        at android.os.Handler.dispatchMessage(Handler.java:95)
        at org.robolectric.shadows.ShadowMessageQueue.dispatchMessage(ShadowMessageQueue.java:165)
        at org.robolectric.shadows.ShadowMessageQueue.access$100(ShadowMessageQueue.java:31)
        at org.robolectric.shadows.ShadowMessageQueue$1.run(ShadowMessageQueue.java:134)
        at org.robolectric.util.Scheduler.runOrQueueRunnable(Scheduler.java:313)
        at org.robolectric.util.Scheduler.postDelayed(Scheduler.java:152)
        at org.robolectric.util.Scheduler.postDelayed(Scheduler.java:141)
        at org.robolectric.shadows.ShadowMessageQueue.enqueueMessage(ShadowMessageQueue.java:147)
        at android.os.MessageQueue.enqueueMessage(MessageQueue.java)
        at android.os.Handler.enqueueMessage(Handler.java:631)
        at android.os.Handler.sendMessageAtTime(Handler.java:600)
        at android.os.Handler.sendMessageDelayed(Handler.java:570)
        at android.os.Handler.post(Handler.java:326)
        at android.support.v4.app.FragmentManagerImpl.scheduleCommit(FragmentManager.java:1877)
        at android.support.v4.app.FragmentManagerImpl.enqueueAction(FragmentManager.java:1860)
        at android.support.v4.app.BackStackRecord.commitInternal(BackStackRecord.java:643)
        at android.support.v4.app.BackStackRecord.commit(BackStackRecord.java:603)
```

It causes problems with RecyclerView and other custom viewgroups that use posted runnables for things - it makes animations infinitely recurse... test authors run into issues like this as part of the development cycle and work around them, but doing depot-wide upgrades of the support library provides a more quantifiable look at how often code breaks due to the unpaused looper behavior. 

2. Maintaining shadows of the support library is unattractive - ideally the support library shouldn't need shadows at all. Unlike the Android framework, the support library has a real, functioning implementation that's available as a jar - and any APIs it has that pose testing problems can change much more easily than a problematic API in the framework. If robolectric is doing a good job of faithfully emulating Android behavior, the support libraries should "just work". 

3. Yes, this is the cost. However, the cost is not so high - simple tests are still pretty simple, but they'd have the added enhancement of being correct. I think it also makes the underlying model of the UI event loop and robolectric's use of the test thread as the UI thread much more obvious to engineers reading robolectric test code.

4. This is already done by runPaused(), and while that's certainly better than doing *everything* inline, it's still the cause of lots of reentrancy errors.

**Comment by virl** [comment]
For completely architecturally correct fix proposal please see #2977

**Comment by charlesmunger** [comment]
@virl That looks like a cool proposal, and I've built similar things (although tightly tied to internal infrastructure). I think it's orthogonal though - how we manage background execution is separate from how we manage UI thread runnables scheduled by the UI thread.

**Comment by virl** [comment]
@charlesmunger There is no need to manage UI thread runnables at all and in that sense they are no different from background execution. 
Current attempt to make UI runnables deterministic and step-by-step testable is fundamentally incorrect, because it breaks encapsulation of tested code.

Any asynchronous testing should be peformed in Xcode's model of XCTestExpectation, when test doesn't know which runnables, on what loopers and in what order will be submitted by tested code.
The only thing test should care about is execution of public interface's callbacks on proper threads.

The test should await completion of these callbacks via CountDownLatch or special Expectation class (which will spin test's runloop for waiting), thus making no assumption or interference in asynchronous execution of tested code — be that execution on main or on background thread.

**Comment by charlesmunger** [comment]
@virl You're correct that in this hypothetical new world there would be no difference from background execution - however, before we can even evaluate how we'd get to the world you're describing, we'd need to get rid of the current inline execution behavior. 

> Current attempt to make UI runnables deterministic and step-by-step testable is fundamentally incorrect, because it breaks encapsulation of tested code.

While it's true that the absence of `runUiThreadTasks` before assertions in a test is relying on some behavior happening inline, that's not actually an unreasonable constraint - for any method with observable side effects, when those side effects are documented to happen is an important part of the method's contract. On android, for example, posting an event to the UI thread can introduce ~32ms of latency just waiting for vsync and double buffering. 

> The test should await completion of these callbacks via CountDownLatch or special Expectation class (which will spin test's runloop for waiting), thus making no assumption or interference in asynchronous execution of tested code — be that execution on main or on background thread.

For many tests, that's correct. If you press a button and want to verify that some text is shown in response, you shouldn't care whether the text is shown immediately or on the next pass through the UI loop. Espresso's view assertions work this way. On the other hand, it's wrong for other tests - commit() and commitNow() are different. If operations are posted to the UI loop, operations scheduled before them may observe invalid state. For this reason, it's important that we make it possible to do assertions without exhausting the looper. 

In addition, people use their own assertion libraries - Truth, regular junit, hamcrest, etc. It's not feasible or reasonable to prevent them from working in Robolectric tests.

**Comment by virl** [comment]
@charlesmunger Proposed solution doesn't prevent assertion libraries from working.

In fact, it is current situation introduces bugs and problems, because it allows test to interfere with UI's intermediate state by checking it on arbitrary looper iteration and not using time or callbacks.

The fact that many tests code rely on current broken Robolectric paradigm is not excuse to not fix it.
It will have to be fixed sooner or later, because more and more app's components become asynchronous and require ability to be tested as blackbox.

It is almost laughable that Robolectric cannot even been used with super-popular mobile DB: #1389 https://github.com/realm/realm-java/issues/904
And it's all because because Robolectric for some reason decided to stop all Loopers in the tested app!

**Comment by charlesmunger** [comment]
@virl I think we're talking about two different things, because the title of my issue was overbroad. My proposal is only about the main thread's looper, not any others - this won't exacerbate any issues with Realm or similar libraries that use their own loopers internally. The UI thread's looper, since it's also the test thread, has unique behavior in robolectric - other loopers don't exhibit the inline execution behavior that it does. 

Do you still have concerns about this proposal as it relates to the main looper, not any others?

**Comment by virl** [comment]
@charlesmunger I didn't have any concerns about your proposal at all.
What I'm saying is that the problem that you encountered with main looper is in fact problems with all loopers in Robolectric.
Your proposal can't exacerbate issues with Realm, because Realm doesn't work in Robolectric due to these looper problems already.

**Comment by charlesmunger** [comment]
@virl Ok, just making sure :-)

I think you should file a new issue about shadow loopers in the background. Ideally we'd just strip out the entire shadow implementation except the parts that are native code, and offer an API to get a callback when they go idle, so that the behavior you want could be implemented.

**Comment by virl** [comment]
@charlesmunger Callback when they go idle is not needed, because loopers should just work the same way they work in Android, by polling thread's message queue.

**Comment by tec27** [comment]
I sort of wonder how much this would actually fix the issues you're trying to solve. IMO the "real" issue with the Looper as it stands is that it handles re-entrant execution of tasks badly (that is, an executing task can end up triggering another run of the Looper while it is executing, which currently causes recursive task execution rather than running things in series). Pausing the looper doesn't really fix that problem cleanly (at least not without other, more significant changes to how some of the async APIs work in Robolectric). And this actually happens *a lot* with current shadow implementations (things like layout inflation tend to cause it).

I have an internal change that makes Robolectric's task execution always be serial (instead of recursive in the re-entrant case), but it ends up being rather more complex than the current implementation, due to how general, low-level, and public the Scheduler/Looper APIs are (and thus leading to a wide array of possible usages). Making that change makes a lot of things work much more similarly to how Android executes them on devices, though (including fragment transactions, layout inflation, layout traversal, etc.)

**Comment by virl** [comment]
@tec27 The real problem is that instead of just implementing Android's Looper behaviour on top of desktop JVM, the Robolectric implements pausable non-reentrant out-of-order global scheduler for all of the loopers.

To fix things, the entire scheduler must be deprecated and ShadowLooper must work exactly as Android's original Looper, just on desktop JVM's Thread.

**Comment by charlesmunger** [comment]
@virl That's not possible for the UI thread, and the behavior of background threads isn't relevant to this proposal. 

@tec27 Isn't your proposal mostly equivalent to calling `runUiThreadTasks()` at the end of every Shadow* method? I'm not in love with that behavior as a final state, but it's certainly a strict improvement over what we have now.

**Comment by virl** [comment]
@charlesmunger UI thread are no different from any other background thread with Looper. Problems with it in Robolectric are really consequence of more general issue of inability to run async code in test and wait for it using timeouts+CountDownLatches in callbacks.

Thinking otherwise will lead to yet another non-working half-solution which will confuse test's mechanics even more and will distance them from instrumentation tests even more.

**Comment by charlesmunger** [comment]
@virl You're right that it's not different, *except in robolectric*. In robolectric, the test thread's `Thread.currentThread() == Looper.getMainLooper().getThread()`. There's no idle state where someone can await new tasks indefinitely. This decision is not likely to change, and either way it's orthogonal to my proposal - this moves the behavior of code under tests closer to the behavior of code on real Android, including instrumentation tests, because it makes tasks on the UI thread run in the correct order.

**Comment by virl** [comment]
@charlesmunger 
> You're right that it's not different, except in robolectric. In robolectric, the test thread's Thread.currentThread() == Looper.getMainLooper().getThread()

It's no problem and in fact, it is completely correct architectural decision, because it removes multithreading/racecondition problems when testing UI code (or code that calls into UI frameworks inside itself).

iOS's XCTest does the same.

> There's no idle state where someone can await new tasks indefinitely.

I already described solution to that in #2977 : Robolectric should introduce Expectation object, waiting on which inside the test will SPIN the Main Looper.

iOS's XCTest does exactly that.

> it's orthogonal to my proposal

It's not orthogonal to your proposal, it's just your proposal is architecturally **incomplete** is just another attempt to quick-patch Robolectric's broken threading architecture.

**Comment by charlesmunger** [comment]
Robolectric already has such an api - ShadowLooper.runUiThreadTasks(). Are you suggesting to call it in a loop until some test condition is satisfied?

> It's not orthogonal to your proposal, it's just your proposal is architecturally incomplete is just another attempt to quick-patch Robolectric's broken threading architecture.
It isn't a quick-patch, and it is an incremental step along the path you're describing. Any changes we make here need to be incremental to allow developers to update their tests in a sane way.

**Comment by virl** [comment]
> Robolectric already has such an api - ShadowLooper.runUiThreadTasks()

No, it does not, because runUiThreadTasks() doesn't spin main looper waiting for condition — instead, it executes Runnables of ALL loopers that it think should be executed "now" and then "freezes" time for them again and returns control.

It is fundamentally incorrect for async code testing, because you cannot depend on execution order **and time** of tested code.

That's main deficiency of your current proposal: you propose to fix execution order **without fixing execution time**.

> It isn't a quick-patch, and it is an incremental step along the path you're describing. Any changes we make here need to be incremental to allow developers to update their tests in a sane way.

No, it is not incremental step, because incremental improvements is impossible in architecturally incorrect system (like Robolectric's thread model) — they will just make it incorrect in different way. So they will break current development guarantees without introducing finally correct behaviour.

**Comment by charlesmunger** [comment]
> No, it does not, because runUiThreadTasks() doesn't spin main looper waiting for condition — instead, it executes Runnables of ALL loopers that it thinks should be executed "now" and then "freezes" time for them again and returns control.

[ShadowLooper.runUiThreadTasks()](https://github.com/robolectric/robolectric/blob/6fa38fc48242cbdcb08b54f4bd5b5a0a8dbee6f2/shadows/framework/src/main/java/org/robolectric/shadows/ShadowLooper.java#L201) only runs tasks on the UI thread, it doesn't touch other loopers. If it's ever going to return control, it has to freeze time for the UI thread looper - the one thread can't be running the test code and posted events. And as for spinning waiting for a condition - there's no point in spinning any further once you've exhausted all your tasks. If some other thread is adding tasks to the UI thread, the correct thing is to coordinate - block the test thread waiting for that thread to queue its task on the UI thread, then `runUiThreadTasks()`. This is essentially the model Espresso uses - rather than spinning waiting for a condition, they wait for the system to go idle, then return to user code so that asserts can run on a steady state. This way you get bounded test runtime - if the condition isn't met, you get a useful assertion message as quickly as possible, rather than a test timeout. 

> It is fundamentally incorrect for async code testing, because you cannot depend on execution order and time of tested code.
That's main deficiency of your current proposal: you propose to fix execution order without fixing execution time.

Normally, you'd be right. But in this specific case, the UI thread's looper is special, because its thread is also the test thread, and Loopers serialize tasks on a single thread. For example, this robolectric code:

```java
doFoo();
new Handler().post(this::doBaz);
doBar();
runUiThreadTasks();
doQux();
```

Is equivalent to this code outside robolectric, say in an instrumentation test:
```java
ConditionVariable condition = new ConditionVariable();
Handler handler = new Handler(Looper.getMainLooper());
handler.post(() -> {
  doFoo();
  handler.post(this::doBaz);
  doBar();
  handler.getLooper().getQueue().addIdleHandler(() -> {
    doQux();
    condition.open();
    return false;
  });
});
condition.block();
```

In all these cases, calling runUiThreadTasks() and then proceeding is equivalent to adding code in an IdleHandler. Both of these code snippets will execute the same way, always; the ordering will 100% always (on real devices, emulators, and robolectric) be Foo, Bar, Baz, Qux. There's no false serialization with the handling of the UI thread. 

> No, it is not incremental step, because incremental improvements is impossible in architecturally incorrect system (like Robolectric's thread model) — they will just make it incorrect in different way. So they will break current development guarantees without introducing finally correct behaviour.

If we make this change, UI thread order now correctly matches the real behavior. It provides all the behavior that you need to implement the model you describe in that other bug - the missing piece is the behavior of the *other* loopers, which isn't affected by this proposal at all, and this proposal doesn't make it worse. The breakages this change (reliance on the inline behavior) will also happen in the model you're describing, so this doesn't cause more churn than if your proposal was accepted and implemented immediately.

**Comment by virl** [comment]
> ShadowLooper.runUiThreadTasks() only runs tasks on the UI thread, it doesn't touch other loopers.

By default yes, but what's important that it doesn't guarantee it if global scheduling is enabled.

> If it's ever going to return control, it has to freeze time for the UI thread looper - the one thread can't be running the test code and posted events.

**That's the problem** : it returns control ARBITRARILY, because any other thread can add task just after it returned control.

> And as for spinning waiting for a condition - there's no point in spinning any further once you've exhausted all your tasks.

You cannot know that you exhausted all your tasks, because it'is  **The Turing Halting Problem**: https://en.wikipedia.org/wiki/Halting_problem

> If some other thread is adding tasks to the UI thread, the correct thing is to coordinate - block the test thread waiting for that thread to queue its task on the UI thread, then runUiThreadTasks().

It's impossible, because you do not have access to that other thread from test code, because that thread is implementation detail of tested class (or some other class it uses).

> This is essentially the model Espresso uses

It doesn't matter, it's broken model.

> rather than spinning waiting for a condition, they wait for the system to go idle
 
 You can't know when system is idle, it's **The Turing Halting Problem**: https://en.wikipedia.org/wiki/Halting_problem
 
 > then return to user code so that asserts can run on a steady state.
 
 Steady state cannot be determined, it's **The Turing Halting Problem**: https://en.wikipedia.org/wiki/Halting_problem
 
 > This way you get bounded test runtime - if the condition isn't met, you get a useful assertion message as quickly as possible, rather than a test timeout.

This leads to impossibility of correct testing of any complext asynchronous system (when tested code uses asynchronous components).

**That's why Realm database still doesn't work in Robolectric**

**That's why Apple's iOS's XCTest uses completely different async testing model**

> But in this specific case, the UI thread's looper is special

It is not special, currently it is broken in the same way as any other Looper in Robolectric.

> because its thread is also the test thread, and Loopers serialize tasks on a single thread.

Any Looper serialize its tasks on single thread.

> Is equivalent to this code outside robolectric, say in an instrumentation test:
> ...
> handler.getLooper().getQueue().addIdleHandler(() ->
> ...

addIdleHandler() is broken for any multithreaded system and should not be used, because it's Turing Halting Problem. Therefore, your entire example is wrong.

> If we make this change, UI thread order now correctly matches the real behavior.

No, it will not: it will not execute scheduled runnables if Thread.sleep(time) called from test code.

**Comment by virl** [comment]
Will your change fix following?

This code is related to main thread looper only, just as you like.
```
@Test
	@Throws
	fun testBrokenRobolectric()
	{
		val latch = CountDownLatch(1)
		val mainHandler = Handler(Looper.getMainLooper())
		mainHandler.postDelayed({
			latch.countDown()
		}, 1000)

		Thread.sleep(2000)

		Assert.assertTrue(latch.await(500, TimeUnit.MILLISECONDS))
	}
```

**Comment by virl** [comment]
Will your proposed change fix the following?
```
@Test
@Throws
fun testBrokenRobolectric()
{
	val latch = CountDownLatch(1)
	val mainHandler = Handler(Looper.getMainLooper())
	mainHandler.postDelayed({
		latch.countDown()
	}, 1000)

	ShadowLooper.unPauseMainLooper()

	Assert.assertTrue(latch.await(500, TimeUnit.MILLISECONDS))
}
```
This code is related to main thread looper only, just as you like.

**Comment by virl** [comment]
> In all these cases, calling runUiThreadTasks() and then proceeding is equivalent to adding code in an IdleHandler

No, it is not:
```
Handler handler = new Handler(Looper.getMainLooper());
		handler.postDelayed(new Runnable()
		{
			@Override
			public void run()
			{
				System.out.println("1");
			}
		}, 1000);

runUiThreadTasks();
System.out.println("2");
```
Result:
2

```
final CountDownLatch latch = new CountDownLatch(1);

Handler handler = new Handler(Looper.getMainLooper());
handler.postDelayed(new Runnable()
{
	@Override
	public void run()
	{
		Log.e("footest", "1");
	}
}, 1000);
handler.getLooper().getQueue().addIdleHandler(new MessageQueue.IdleHandler()
{
	@Override
	public boolean queueIdle()
	{
		Log.e("footest", "2");
		latch.countDown();
		return false;
	}
});

latch.await();
```
Result:
1
2

Therefore, your entire example is wrong.

**Comment by virl** [comment]
> In all these cases, calling runUiThreadTasks() and then proceeding is equivalent to adding code in an IdleHandler

No, it is not:
```
Handler handler = new Handler(Looper.getMainLooper());

handler.post(new Runnable()
{
	@Override
	public void run()
	{
		System.out.println("1");
	}
});

handler.postDelayed(new Runnable()
{
	@Override
	public void run()
	{
		System.out.println("2");
	}
}, 1000);

runUiThreadTasks();
System.out.println("3");
```
Result:
1
3

```
final CountDownLatch latch = new CountDownLatch(1);

Handler handler = new Handler(Looper.getMainLooper());

handler.post(new Runnable()
{
	@Override
	public void run()
	{
		Log.e("footest", "1");
	}
});

handler.postDelayed(new Runnable()
{
	@Override
	public void run()
	{
		Log.e("footest", "2");
	}
}, 1000);

Thread.sleep(new Random().nextInt(2000));

handler.getLooper().getQueue().addIdleHandler(new MessageQueue.IdleHandler()
{
	@Override
	public boolean queueIdle()
	{
		Log.e("footest", "3");
		latch.countDown();
		return false;
	}
});

latch.await();
```
Result:
Unknown

Therefore, your entire example is wrong.

**Comment by JakeWharton** [comment]
Big :+1: to the proposal. 100% of my tests that actually care about the fact that they're on main loop start with `@Before { pauseMainLooper() }`.

**Comment by virl** [comment]
@charlesmunger Jack with his optimistic comment reminded me that your proposal is honest attempt at incremental improvement.
So please don't take to heart my criticisms of it: I only want to show that this entire problem is already solved on other mobile platform (iOS) in completely correct way, so maybe it will be saving of energy to implement final solution right away.

**Comment by charlesmunger** [comment]
As an update to this proposal - an experiment of setting the looper to paused by default breaks about 8% of Robolectric tests in Google's monorepo. That's a much smaller portion than I'd anticipated, and with an incremental migration is very tractable for a whitelist approach to migration.  @jongerrish

**Comment by jongerrish** [comment]
https://github.com/robolectric/robolectric/issues/2851

**Comment by jongerrish** [comment]
https://github.com/robolectric/robolectric/issues/2851

**Comment by AstralStorm** [comment]
Here's a related question, now with the new PAUSED Looper, how do we run all loopers (including hidden unknown loopers in HandlerThreads) in a specified order, or at all?

**Comment by copybara-robolectric** [comment]
I assume by hidden Loopers you mean ones that your test cannot easily get a
reference too.

We had talked about adding apis for this kind of situation, but we weren't
sure of the need. Both shadow looper implementation variants do hold a
reference to all the active loopers, so having APIs for this sort of thing
would be easy to add.

Would something like ShadowLooper#getAllLoopers and/or
ShadowLooper#pauseAll, shadowLooper#idleAll solve this concern?

**Comment by AstralStorm** [comment]
For now, ShadowSystemClock.advanceBy(...) works for all of Loopers. But it's relatively limited.

**Comment by brettchabot** [comment]
tracking here https://github.com/robolectric/robolectric/issues/5098

**Comment by utzcoz** [comment]
Duplicated to https://github.com/robolectric/robolectric/issues/5098.
